### DISCUSSION TITLE
A question about a 'while True' loop

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#13541] [General]
Hi all,

I've been exploring MQTT and I came across some code that puzzles me.
The code is here: https://www.emqx.com/en/blog/micro-python-mqtt-tutorial-based-on-raspberry-pi#subscribe
(See below)

The part I don't understand is this:
````
while True:
        if True:
            client.wait_msg()
        else:
            client.check_msg()
            time.sleep(1)
````

How can the 'else' condition ever be met?
While True would mean that 'True' will always be true right?
And what is even being evaluated here, what should be true?

I'm sorry for sounding confused but true can't be false right?
My mind is stuck in a loop now :)

If someone can explain, please do.

Thanks.

========================
Complete code (without wifi connect)
```
# sub.py
import time
from umqtt.simple import MQTTClient

SERVER="broker.emqx.io"
ClientID = f'raspberry-sub-{time.time_ns()}'
user = "emqx"
password = "public"
topic = "raspberry/mqtt"
msg = b'{"msg":"hello"}'

def sub(topic, msg):
    print('received message %s on topic %s' % (msg, topic))

def main(server=SERVER):
    client = MQTTClient(ClientID, server, 1883, user, password)
    client.set_callback(sub)
    client.connect()
    print('Connected to MQTT Broker "%s"' % (server))
    client.subscribe(topic)
    while True:
        if True:
            client.wait_msg()
        else:
            client.check_msg()
            time.sleep(1)

if __name__ == "__main__":
    main()
```

**Comment by davefes** [comment]
I wonder if:
```py
if True:
    client.wait_msg()
```
is an idiom, somehow evaluating the return value  of `client.wait_msg()`

**Comment by shariltumin** [comment]
The code snippet shown is confusing because the comments have been removed. The original is from "lib/micropython-lib/micropython/umqtt.simple/example_sub.py":

```py
...
    while True:
        if True:
            # Blocking wait for message
            c.wait_msg()
        else:
            # Non-blocking wait for message
            c.check_msg()
            # Then need to sleep to avoid 100% CPU usage (in a real
            # app other useful actions would be performed instead)
            time.sleep(1)
...
```

The example forces the use of "blocking wait" and does nothing else. 

To do other processing, the "non-blocking wait" is provided. Have a look at "lib/micropython-lib/micropython/umqtt.simple/umqtt/simple.py" to see how this is done.

The callback is called in wait_msg(). A very clever way of using sock.setblocking(); **False** in check_msg() and **True** in wait_msg(), after sock.read(1). This does a "non-blocking wait" if no data is available and processes the incoming message if data is available.

**Reply by MrTinkerman** [reply]
Thanks for the info!
